File: python/2015/day03a/day03a.py
#!/usr/bin/env python3

import sys

def usage():
    progname = sys.argv[0]
    print(f'usage: {progname} <input file>')
    sys.exit(1)

def process(filename):
    positions = set()
    santa = (0, 0)
    positions.add(santa)
    with open(filename, 'r') as infile:
        for line in infile:
            for ch in line:
                if ch == '^':
                    santa = (santa[0], santa[1] + 1)
                elif ch == 'v':
                    santa = (santa[0], santa[1] - 1)
                elif ch == '<':
                    santa = (santa[0] - 1, santa[1])
                elif ch == '>':
                    santa = (santa[0] + 1, santa[1])
                positions.add(santa)
    return len(positions)

def main():
    if len(sys.argv) < 2:
        usage()
    filename = sys.argv[1]
    result = process(filename)
    print(f'result = {result}')

if __name__ == '__main__':
    main()

File: perl/2015/day03a/day03a.pl
#!/usr/bin/env perl

use strict;
use warnings;

sub usage {
    my $progname = $0;
    print "usage: $progname <input file>\n";
    exit 1;
}

sub process {
    my ($filename) = @_;
    my %positions;
    my $santa = [0, 0];
    my $position = join ',', @$santa;
    $positions{$position} = 1;

    open ( my $infile, '<', $filename ) or die "cannot open file: $filename: $!";
    while ( my $line = <$infile> ) {
        foreach my $ch ( split //, $line ) {
            if ( $ch eq '^' ) {
                $santa->[1] += 1;
            } elsif ( $ch eq 'v' ) {
                $santa->[1] -= 1;
            } elsif ( $ch eq '<' ) {
                $santa->[0] -= 1;
            } elsif ( $ch eq '>' ) {
                $santa->[0] += 1;
            }
            $position = join ',', @$santa;
            $positions{$position} = 1;
        }
    }
    close($infile);

    return scalar keys %positions;
}

sub main {
    if ( @ARGV < 1 ) {
        usage();
    }
    my $filename = $ARGV[0];
    my $result = process($filename);
    print "result = $result\n";
}

main();
